rules/firestore.json

```diff
@@ -35,50 +36,188 @@ service cloud.firestore {
-            // Anybody can read the fields of a user
+            // Only an authenticated account can create its own user, and they cannot directly be verified
+            allow create: if isOwner(userId) && request.resource.data.verified == false
+            // Only the user can change its fields
+            // Moreover, the username cannot be changed once set
+            // And the verified field can only be changed to true if the user has a username and has its email verified
+            allow update: if isOwner(userId) && doesNotModifyUsernameIfSet() && onlyModifiesVerifiedFieldIfUserIsVerified();
+            // Anybody can read the fields of a user.
+            // This is because before user creation, we need to check if the username is free.
             allow read: if true;
         }
-        match /parties/{id} {
-            allow write, read: if userVerified();
-        }
-        match /joiners/{id} {
-            allow write, read: if userVerified();
+        function isOwner(userId) {
+            return request != null &&
+                request.auth != null &&
+                request.auth.uid == userId;
         }
-        function userVerified() {
-            return isAuthentified() &&
+        function doesNotModifyUsernameIfSet() {
+            let usernameNotYetSet = hasField(resource.data, 'username') == false;
+            let usernameBeingSet = hasField(request.resource.data, 'username');
+            // It sometimes happens that firebase sends the username as part as updates, even though it is not actually updated. Hence, we accept usernames as part of updates, only if it doesn't change.
+            let doesNotChange = hasField(resource.data, 'username') && hasField(request.resource.data, 'username') && resource.data['username'] == request.resource.data['username'];
+            return usernameNotYetSet || usernameBeingSet == false || doesNotChange;
+        }
+        function onlyModifiesVerifiedFieldIfUserIsVerified() {
+            let verifiedKeptSameValue = (resource.data.verified == request.resource.data.verified);
+            let verifiedSetToTrueAfterVerification =  (resource.data.verified == false && userIsVerified() && resource.data.username != null);
+            return verifiedKeptSameValue || verifiedSetToTrueAfterVerification
+        }
+        match /parts/{id} {
+            // Anybody can create a part
+            allow create: if userIsVerified();
+            // The owner of a part can delete it, if it has not started
+            allow delete: if userIsVerified() &&
+                             creatorOf(id).id == request.auth.uid &&
+                             partHasNotStarted(resource.data);
+            // Anyone can delete a part if the owner has timed out and the part has not started
+            allow delete: if userIsVerified() &&
+                             hasTimedOut(creatorOf(id)) &&
+                             partHasNotStarted(resource.data);
```

Comment:
ah, on peut faire plusieur "allow action" de suite, pas mal plutôt que dupliquer les et/ou !

Reply:
allow plusieurs fois c'est pour les ou, pas pour les et hélas !

Reply:
Pour les OU c'est très bien !
allow: creatorCreating()
allow: candidateArriving()
allow: candidateLeaving()
allow: creatorKickingCandidate()
allow: chosenOpponentAccepting()
C'bien des or, c'est très bon !

Ce pattern serait pas mal, lisible en plus non ?

Reply:
oui oui c'est c'qui est fait désormais non ?

Reply:
pas entièrement mais j'suis très satisfait du ticket, CONTINUONS (on peut ticketter un "make it a bit more perfect" ticket :D )

Reply:
je pense que ticketter du concret serait mieux :D Et du plus petit aussi !

Reply:
Adapt all security check to pattern "allow method if: isRoleDoingAction()" ?

Reply:
TODO mais oui très mieux